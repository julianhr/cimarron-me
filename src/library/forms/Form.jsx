import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const FormTag = styled.form`
  display: flex;
  flex-direction: column;

  div.field {
    margin-bottom: 15px;
  }

  label {
    padding: 0 10px 4px;
    width: 100%;
    display: flex;
    color: ${props => props.theme.colors.primary.dark};
    font-size: 1.1em;
  }

  input, textarea {
    width: 100%;
    margin-bottom: 2px;

    :focus {
      outline: none;
    }
  }

  input {
    padding: 10px 12px;
  }

  textarea {
    padding: 12px;
  }

  span {
    color: ${props => props.theme.colors.error};
    font-size: 14px;
    display: flex;
    height: 16px;
  }

  button[type=submit] {
    margin-top: 20px;
  }
`

function Form({ children, formStyle, onSubmit }) {
  return (
    <FormTag
      css={formStyle}
      method={'POST'}
      onSubmit={onSubmit}
      onKeyPress={handleFormOnKeyPress}
    >
      {children}
    </FormTag>
  )
}

export function handleFormOnKeyPress(event) {
  if (event.key !== 'Enter') { return null }

  // allow user to press Enter inside a textarea tag
  // insert a line break instead of submitting the form
  if (event.target.type === 'textarea') {
    event.stopPropagation()
  }
}

Form.propTypes = {
  formStyle: PropTypes.object,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ]).isRequired,
}

Form.defaultProps = {
  formStyle: {},
}

export default Form
