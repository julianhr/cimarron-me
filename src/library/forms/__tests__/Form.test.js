import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'
import emotionSerializer from 'jest-emotion'

import Form, { handleFormOnKeyPress } from '../Form'
import MockApp from '~/__tests__/__mocks__/MockApp'


const testProps = {
  formStyle: { color: 'green' },
  onSubmit: () => {},
}

const createApp = (props) => create(
  <MockApp>
    <Form {...props}>
      <input type="text" defaultValue="test value"/>
    </Form>
  </MockApp>
)

const mountApp = (props) => mount(
  <MockApp>
    <Form {...props}>
      <input type="text" defaultValue="test value"/>
    </Form>
  </MockApp>
)

describe('Form', () => {
  describe('props', () => {
    test('formStyle is not required', () => {
      const props = {...testProps}
      delete props.formStyle
      expect(() => createApp(props)).not.toThrow()
    })

    it('incorporates formStyle if passed', () => {
      const props = {...testProps}
      props.formStyle = { color: 'salmon' }
      expect.addSnapshotSerializer(emotionSerializer)
      expect(createApp(props)).toMatchSnapshot()
    })

    test('onSubmit is required', () => {
      const props = {...testProps}
      delete props.onSubmit
      expect(() => createApp(props)).toThrow()
    })

    it('calls onSubmit function when form is submitted', () => {
      const spy = jest.spyOn(testProps, 'onSubmit')
      const wrapper = mountApp(testProps)
      wrapper.find('form').simulate('submit')
      expect(spy).toHaveBeenCalledTimes(1)
      wrapper.unmount()
    })

    test('children is required', () => {
      expect(() => create(<MockApp><Form {...testProps} /></MockApp>)).toThrow()
    })

    test('children can be an element', () => {
      expect(() => create(
        <MockApp>
          <Form {...testProps}>
            <input type="text" defaultValue="test value"/>
          </Form>
        </MockApp>)
      ).not.toThrow()
    })

    test('children can be a fragment', () => {
      expect(() => create(
        <MockApp>
          <Form {...testProps}>
            <>
              <input type="text" defaultValue="test value 1"/>
              <input type="text" defaultValue="test value 1"/>
            </>
          </Form>
        </MockApp>)
      ).not.toThrow()
    })
  })

  describe('helpers', () => {
    describe('handleFormOnKeyPress', () => {
      it('returns null if key pressed is not Enter', () => {
        const fakeEvent = { key: 'a', target: { type: 'textarea' } }
        expect(handleFormOnKeyPress(fakeEvent)).toBe(null)
      })

      it('does nothing if element is not a textarea', () => {
        const spy = jest.fn()
        const fakeEvent = { key: 'Enter', target: { type: 'input' }, stopPropagation: spy }
        handleFormOnKeyPress(fakeEvent)
        expect(spy).not.toHaveBeenCalled()
      })

      it('calls stopPropagation if element is textarea', () => {
        const spy = jest.fn()
        const fakeEvent = { key: 'Enter', target: { type: 'textarea' }, stopPropagation: spy }
        handleFormOnKeyPress(fakeEvent)
        expect(spy).toHaveBeenCalled()
      })
    })
  })
})
