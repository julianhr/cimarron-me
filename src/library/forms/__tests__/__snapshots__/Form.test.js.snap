// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form props incorporates formStyle if passed 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: salmon;
}

.emotion-0 div.field {
  margin-bottom: 15px;
}

.emotion-0 label {
  padding: 0 10px 4px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #9B1518;
  font-size: 1.1em;
}

.emotion-0 input,
.emotion-0 textarea {
  width: 100%;
  margin-bottom: 2px;
}

.emotion-0 input:focus,
.emotion-0 textarea:focus {
  outline: none;
}

.emotion-0 input {
  padding: 10px 12px;
}

.emotion-0 textarea {
  padding: 12px;
}

.emotion-0 span {
  color: #d400ff;
  font-size: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 16px;
}

.emotion-0 button[type=submit] {
  margin-top: 20px;
}

<form
  className="emotion-0 emotion-1"
  method="POST"
  onKeyPress={[Function]}
  onSubmit={[Function]}
>
  <input
    defaultValue="test value"
    type="text"
  />
</form>
`;
